Merge GpuMgr, L0 allocators and GPU buffer entities into jit-engine

dchigarev opened the pull request:
What this PR does:
1. Adds `GpuMgr` interface that's implemented by `CudaMgr` and `L0Manager`. All of the interface code was copied from `gpu_support` branch, for except of adding two new interface methods for broader compatibility:
    - `int getDeviceCount() const`
    - `GpuMgrName getName() const`
2. Aligns `L0Manager` with the `GpuMgr` interface (implement these two new methods)
3. Adds `L0Allocator`, `GpuL0Buffer`, and `GpuL0BufferMgr` L0 implementations for `DeviceAllocator`, `Buffer`, and `BufferMgr` accordingly (implementations were copied from gpu_support branch with no changes)
4. Replaces `CudaMgr` with an abstract `GpuMgr` in `CpuBuffer`, `CpuBufferMgr`, `TieredCpuBuffer`.
5. Allows passing multiple gpu managers into `DataMgr`:
    - Now `DataMgr` accepts a vector of unique_ptrs pointing to an abstract `GpuMgr` instead of a single pointer to `CudaMgr`
    - An actual handling of multiple managers is not implemented yet, the code raises a `CHECK(false)` if it's required to use a gpu manager when multiple instances of them are specified in the `DataMgr`
    - There are still a couple of places inside `DataMgr` where it's required to do manager-specific branching
6. Adds a sanity test for `DataMgr` with specified `L0Manager` (`Tests/DataMgrWithL0.cpp`)

dchigarev commented on DataMgr/DataMgr.h:
```diff
@@ -45,10 +48,6 @@ class FileBuffer;
 class GlobalFileMgr;
 }  // namespace File_Namespace
 
-namespace CudaMgr_Namespace {
-class CudaMgr;
-}
-
```
The reason for replacing this with the actual CudaMgr include is that we need a complete type in order to dynamic cast inside `getCudaMgr()`

dchigarev commented on DataMgr/DataMgr.h:
```diff
@@ -173,7 +172,7 @@ class DataMgr {
       const std::string& dataDir,
       const SystemParameters& system_parameters,
-      std::unique_ptr<CudaMgr_Namespace::CudaMgr> cudaMgr,
+      std::vector<std::unique_ptr<GpuMgr>>&& gpuMgrs,
```
Not sure if this a good approach, maybe we should use something like `unique_ptr<std::vector<...>>` instead of `std::vector<...>&&`?

dchigarev commented on DataMgr/DataMgr.cpp:
```diff
@@ -248,39 +258,74 @@ void DataMgr::populateMgrs(const SystemParameters& system_parameters,
+          deviceMemSize = getCudaMgr()->getDeviceProperties(gpuNum)->globalMem;
+          break;
+        case GpuMgrName::L0:
+          deviceMemSize = 1024 * 4 * 1024;  // 4MB for now
+          page_size = 4096UL;
```
these values were copied from [`gpu_support` branch](https://github.com/intel-ai/omniscidb/blob/2e1a126a384889583c8670e19d5c35b7ff1428be/DataMgr/DataMgr.cpp#L263-L272)

kurapov-peter replied:
Can't it be virtual to avoid the switch statement btw?

dchigarev commented on L0Mgr/L0Mgr.h:
```diff
@@ -202,11 +203,13 @@ class L0Manager {
 
   void synchronizeDevices() const;
+  GpuMgrName getName() const override { return GpuMgrName::L0; }
+  int getDeviceCount() const override { return drivers_.size() ? drivers_[0]->devices().size() : 0; }
```
Not sure if this is valid, can we assume that all of the devices can be observed from `drivers_[0]`?

kurapov-peter replied:
We can not. This will work for a single driver only. OK for testing, not OK for real cases.

alexbaden commented on DataMgr/BufferMgr/CpuBufferMgr/CpuBuffer.cpp:
```diff
@@ -15,22 +15,22 @@
                      BufferList::iterator segment_iter,
                      const int device_id,
-                     CudaMgr_Namespace::CudaMgr* cuda_mgr,
+                     GpuMgr* gpu_mgr,
```
A general comment -- does this ownership model make sense going forward? What does the flow look like if I have a CPU Buffer and want to transfer it to GPU now?

dchigarev replied:
> What does the flow look like if I have a CPU Buffer and want to transfer it to GPU now?

1. If you want to convert `CpuBuffer` object into `GpuCuda/L0Buffer` then the flow is the following:

    There is no explicit convertion from `CpuBuffer` to `GpuBuffer`, so you have to construct a new GPU buffer by yourself   (e.g. using `GpuBufferMgr`) and write the data from `CpuBuffer` in it:

    ```cpp
    // `.alloc()` returns an AbstractBuffer, so doing a cast
    CpuBuffer* cpu_buff = (CpuBuffer*)CpuBufferMgr.alloc(n_bytes);
    cpu_buff.write(...); // write some data to the buffer

    // Transfering data from CpuBuffer to GpuL0Buffer:
    // 1. Create a gpu buffer mgr to allocate gpu buffer
    GpuL0BufferMgr gpu_buff_mgr{
                /*mgr params*/
                            l0_mgr,
                /*mgr params*/
            };
    // 2. Allocate GPU buffer
    GpuL0Buffer* gpu_buff = (GpuBuffer*)gpu_buff_mgr.alloc(n_bytes);
    // 3. Copy the data from CPU to GPU buffer
    gpu_buff->write(
        /*src=*/cpu_buff->getMemoryPtr(),
        /*nbytes=*/n_bytes,
        /*offset=*/0,
        /*src_buffer_type=*/MemoryLevel.CPU_LEVEL
    );
    ```

2. If you already have a pointer to allocated memory on GPU and just want to transfer data in it, you can do the required copy via `CpuBuffer` by passing appropriate parameters:
    ```cpp
    // `.alloc()` returns an AbstractBuffer, so doing a cast
    CpuBuffer* cpu_buff = (CpuBuffer*)CpuBufferMgr.alloc(n_bytes);
    cpu_buff.write(...); // write some data to the buffer

    int8_t* gpu_ptr = getGpuMemorySomehow();
    cpu_buff.read(
        gpu_ptr,
        n_bytes,
        /*offset=*/0,
        /*dst_buffer_type*/MemoryLevel::GPU_LEVEL // this parameters triggers transfer to GPU
    );
    ```

So the reason of having a `GpuMgr` inside `CpuBuffer` is to be able to do the second scenario of Cpu->Gpu transfering, which looks unsafe IMO (as it requires to operate with raw pointers). If we would like to omit/restrict the second use case, then we can safely remove this `GpuMgr` referencing in `CpuBuffer` I think.

alexbaden replied:
I agree, something to look at in our refactorings @ienkovich

alexbaden commented on DataMgr/DataMgr.cpp:
```diff
@@ -45,23 +46,28 @@ namespace Data_Namespace {
-      hasGpus_ = true;
-    } else {
-      LOG(ERROR) << "CudaMgr instance is invalid, falling back to CPU-only mode.";
+    gpuMgrs_.reserve(gpuMgrs.size());
```
Is it possible to have mixed CUDA + L0 GPU mgrs now? Is there a test for this if so?

Edit: Nevermind, I saw the comment below. Why not make the mgr a single member for now, then, and not a vector?

dchigarev replied:
We would potentially want to offload computations to various type of devices, which would make us to operate with multiple GPU managers inside `DataMgr`. For now, there's no such ability, so we can use a single manager indeed instead of a vector, but I thought it makes sense to have the vector now, in case we would like to extend `DataMgr` to work with multiple managers in the future.

alexbaden replied:
I think if that's the case we will want to have a map or something more explicit -- a `Device : Mgr` mapping for example since you probably can't always index devices across different drivers. So, I would drop the vector for now unless @kurapov-peter is making use of it in his L0 branches.

kurapov-peter replied:
No, as Dmitry says, vector was introduces only to minimize code rewriting in the future when we'd want to run on both platforms simultaneously. A map could be better, I agree.

dchigarev replied:
Replaced vector with a map

alexbaden reviewed (commented):
Looks fine. Let's add an inheritance diagram for the newly introduced Mgrs, both GPU mgrs and the new buffer/buffermgr variants.

alexbaden commented:
> @alexbaden, I've made a diagram showing relations between the Mgrs/Buffers and DataMgr, however I have no idea where and how to put it...

Thanks! Docs source is fine, I think we should get the docs up and running in github pages or something similar.

kurapov-peter commented on Tests/DataMgrWithL0Test.cpp:
```diff
@@ -0,0 +1,45 @@
+#include "DataMgr/DataMgr.h"
+#include "TestHelpers.h"
+
+TEST(DataMgrWithL0, SanityTest) {
```
Does it work if ENABLE_L0 is off?

dchigarev replied:
It could work but it won't as the test is enabling [conditionaly depending on ENABLE_L0](https://github.com/dchigarev/omniscidb/blob/85f343c2685f58e639ef0e98e7b33e30cf521733/Tests/CMakeLists.txt#L120-L122)

kurapov-peter replied:
Ok, that will do

dchigarev commented on DataMgr/Allocators/L0Allocator.h:
```diff
@@ -0,0 +1,66 @@
+class DataMgr;
+}  // namespace Data_Namespace
+
+class L0Allocator : public DeviceAllocator {
```
GPU manager specific implementations of `DeviceAllocator`, `Buffer`, and `BufferMgr` were removed. Now there're common `GpuAllocator`, `GpuBuffer`, and `GpuBufferMgr` implementations using an abstract `GpuMgr`.

dchigarev commented on DataMgr/Allocators/L0Allocator.h:
```diff
@@ -0,0 +1,66 @@
+  std::vector<Data_Namespace::AbstractBuffer*> owned_buffers_;
+
+  Data_Namespace::DataMgr* data_mgr_;
+  int device_id_;
```
L0Allocator was removed, so outdated

dchigarev commented:
Last updates for the PR:

1. Got rid of gpu manager specific implementations of DeviceAllocator, Buffer, and BufferMgr, now there're implementations that use an abstract GpuMgr underneath instead of L0Mgr or CudaMgr (`DataMgr/Allocators/GpuAllocator`, `DataMgr/BufferMgr/GpuBufferMgr/..`), this had required to expand `GpuMgr` interface by adding the following methods:
    - zeroDeviceMem
    - setDeviceMem
    - allocateDeviceMem
    - freeDeviceMem
    - synchronizeDevices
    - setContext (proposed to remove as the method seems to make no sense, see in-code comment)
2. Replaced `std::vector` with `std::map` as a GpuMgrs container inside DataMgr.
3. Added `DataMgr::setContext(GpuMgrName)` method that's suppose to switch gpu managers for the DataMgr.

dchigarev commented on DataMgr/BufferMgr/GpuBufferMgr/GpuBufferMgr.cpp:
```diff
@@ -14,17 +14,19 @@
-
+#include "DataMgr/BufferMgr/GpuBufferMgr/GpuBufferMgr.h"
+#ifdef HAVE_CUDA
 #include "CudaMgr/CudaMgr.h"
-#include "DataMgr/BufferMgr/GpuCudaBufferMgr/GpuCudaBuffer.h"
+#endif
```
the import of `CudaMgr.h` is still needed in order to properly handle cuda-specific exceptions in GpuBufferMgr destructor

dchigarev commented on L0Mgr/CMakeLists.txt:
```diff
@@ -5,6 +5,5 @@ else()
 endif()
 
 add_library(L0Mgr ${L0_SOURCES})
-target_include_directories(L0Mgr PRIVATE ${LevelZero_INCLUDE_DIRS})
```
including of LevelZero headers was moved to the [root CMake](https://github.com/dchigarev/omniscidb/blob/85f343c2685f58e639ef0e98e7b33e30cf521733/CMakeLists.txt#L139)

dchigarev commented on L0Mgr/L0Exception.h:
```diff
@@ -14,12 +14,12 @@
 using L0result = int;
 
-class L0Exception : public std::exception {
+class L0Exception : public std::runtime_error {
```
L0Exception has to be an instance of runtime_error: when a BufferMgr allocates a slab, it queries an underlying manager for the maximum available memory to allocate by doing something like:
```cpp
while (a_lot_of_bytes > 0) {
    try {
        allocateMemory(a_lot_of_bytes);
    } catch (std::runtime_error&) {
        a_lot_of_bytes -= step;
    }
}
```
[[actual code]](https://github.com/dchigarev/omniscidb/blob/85f343c2685f58e639ef0e98e7b33e30cf521733/DataMgr/BufferMgr/BufferMgr.cpp#L299-L346)
meaning that exceptions a manager raises has to be derivatives of runtime_error, otherwise slab allocation function works incorrect

kurapov-peter commented on QueryRunner/QueryRunner.cpp:
```diff
@@ -193,10 +195,14 @@ QueryRunner::QueryRunner(const char* db_path,
-    cuda_mgr = std::make_unique<CudaMgr_Namespace::CudaMgr>(-1, 0);
+    gpuMgrs[GpuMgrName::CUDA] = std::make_unique<CudaMgr_Namespace::CudaMgr>(-1, 0);
+  }
+#elseif HAVE_L0
```
`#elseif` is not defined by cpp standard, please use `#elif`

kurapov-peter reviewed (commented):
Minor comments, LGTM

kurapov-peter commented:
I think this is ready, @ienkovich do you want to merge it before or after #351 ?

ienkovich commented:
> I think this is ready, @ienkovich do you want to merge it before or after #351 ?

I've merged #351 already, so this one might require a rebase.

kurapov-peter reviewed (approved):
Works for me
